### PR TITLE
[#172376369] Add password login for reviewers and integration tests

### PR DESCRIPTION
**Pull request by BurnedMarshal**

### CODE REVIEW

File: api_public.yaml [hunk 1]
```diff
@@ -25,7 +25,36 @@ paths:
               version: "0.0.1"
               min_app_version: "0.0.0"
               min_app_version_pagopa: "0.0.0"
+  "password-login":
```
Comment (gunzip) [inline api_public.yaml:28]:
```suggestion
  "test-login":
```

File: api_public.yaml [hunk 2]
```diff
@@ -25,7 +25,36 @@ paths:
...
+    x-swagger-router-controller: AuthenticationController
+    post:
+      operationId: acs
```
Comment (gunzip) [inline api_public.yaml:31]:
```suggestion
      operationId: testLogin
```

File: src/strategies/localStrategy.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  validPassword: string
+): passport.Strategy =>
+  new Strategy((username, password, done) => {
+    if (
+      validUsernameList.some(_ => _ === username) &&
+      validPassword === password
+    ) {
+      // Fake test user for password based logins
+      const testUser = {
+        familyName: "User",
```
Comment (gunzip) [inline src/strategies/localStrategy.ts:16]:
```suggestion
        familyName: "Rossi",
```

File: src/strategies/localStrategy.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        fiscalNumber: username as FiscalCode,
+        getAssertionXml: () => "",
+        issuer: "IO",
+        name: "Test"
```
Comment (gunzip) [inline src/strategies/localStrategy.ts:20]:
```suggestion
        name: "Mario"
```

File: src/app.ts [hunk 5]
```diff
@@ -133,6 +136,12 @@ export function newApp(
   // Add the strategy to authenticate proxy clients.
   passport.use("bearer.wallet", bearerWalletTokenStrategy(SESSION_STORAGE));
 
+  // Add the strategy for reviewers login.
+  passport.use(
+    "local",
+    localStrategy(TEST_LOGIN_FISCAL_CODE, TEST_LOGIN_PASSWORD)
```
Comment (gunzip) [inline src/app.ts:142]:
```suggestion
    localStrategy(TEST_LOGIN_FISCAL_CODES, TEST_LOGIN_PASSWORD)
```

File: src/config.ts [hunk 6]
```diff
@@ -297,3 +298,13 @@ export const BARCODE_ALGORITHM = NonEmptyString.decode(
 
 // Application insights sampling percentage
 export const DEFAULT_APPINSIGHTS_SAMPLING_PERCENTAGE = 20;
+
+// Reviewer static login params
+export const TEST_LOGIN_FISCAL_CODE: ReadonlyArray<FiscalCode> = getRequiredENVVar(
```
Comment (gunzip) [inline src/config.ts:303]:
```suggestion
export const TEST_LOGIN_FISCAL_CODES: ReadonlyArray<FiscalCode> = getRequiredENVVar(
```

File: src/config.ts [hunk 7]
```diff
@@ -297,3 +298,13 @@ export const BARCODE_ALGORITHM = NonEmptyString.decode(
...
+  "TEST_LOGIN_FISCAL_CODE"
```
Comment (gunzip) [inline src/config.ts:304]:
```suggestion
  "TEST_LOGIN_FISCAL_CODES"
```

File: src/config.ts (hunk 6 above)
Comment (gunzip) [inline src/config.ts:303]: can we make these optional ?

File: api_public.yaml (hunk 1 above)
Comment (gunzip) [reply api_public.yaml:28]: it looks like api does not lint:

```
ERRORS
  #/paths: Additional properties not allowed: password-login

```

File: src/config.ts (hunk 6 above)
Comment (BurnedMarshal) [reply src/config.ts:303]: Optional with a default value?

File: src/config.ts (hunk 7 above)
Comment (BurnedMarshal) [reply src/config.ts]: Fixed in 48b864c

File: src/config.ts (hunk 6 above)
Comment (BurnedMarshal) [reply src/config.ts]: Fixed in 48b864c

File: src/app.ts (hunk 5 above)
Comment (BurnedMarshal) [reply src/app.ts]: Fixed in 48b864c

File: src/strategies/localStrategy.ts (hunk 4 above)
Comment (BurnedMarshal) [reply src/strategies/localStrategy.ts]: Fixed in 48b864c

File: src/strategies/localStrategy.ts (hunk 3 above)
Comment (BurnedMarshal) [reply src/strategies/localStrategy.ts]: Fixed in 48b864c

File: api_public.yaml (hunk 1 above)
Comment (BurnedMarshal) [reply api_public.yaml]: Fixed in 48b864c
I've also changed the path including missing base path

File: README.md [hunk 8]
```diff
@@ -174,6 +174,8 @@ Those are all Environment variables needed by the application:
 | APICACHE_DEBUG                         | When is `true` enable the apicache debug mode                                     | boolean |
 | GITHUB_TOKEN                           | The value of your Github Api Key, used in build phase                             | string |
 | FETCH_KEEPALIVE_ENABLED                | When is `true` enables `keepalive` agent in the API client (defaults to `false`)  | boolean |
+| TEST_LOGIN_FISCAL_CODE                 | Enabled username for password based login (coma separated Fiscal Code)            | string |
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply README.md]: Fixed in 48b864c

File: .env.example [hunk 9]
```diff
@@ -30,3 +30,5 @@ SHUTDOWN_TIMEOUT_MILLIS=30000
 GITHUB_TOKEN=<mygithubtoken>
 SPID_LOG_STORAGE_CONNECTION_STRING=foobar
 SPID_LOG_QUEUE_NAME=spidqueuename
+TEST_LOGIN_FISCAL_CODE=AAABBB01C02D345Z,EEEFFF01C02H345Z
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply .env.example]: Fixed in 48b864c

File: src/config.ts (hunk 6 above)
Comment (gunzip) [reply src/config.ts]: nope, without. endpoint is a no-op in case they're missing

Comment (BurnedMarshal) [reply src/config.ts]: @gunzip how do you suggest to implement a no-op response? I've not found a suggestion on [Zalando guidelines](https://opensource.zalando.com/restful-api-guidelines/#http-status-codes-and-errors), maybe a status `304` response, but can't be applied on `POST` (following Zalando guidelines). Maybe we can return a simple Unauthorized.

Comment (gunzip) [reply src/config.ts]: yes it's ok to return 401

Comment (BurnedMarshal) [reply src/config.ts]: fixed in 48b864c

Comment (BurnedMarshal) [reply src/config.ts]: Sorry, fixed in b11cff3

**Pull request merged**